### DISCUSSION TITLE
Multi-Node Single-Drive

### DISCUSSION THREAD

**Discussion by leandroleal** [minio/minio#16047] [Q&A] [answered]
I have 2 new servers (`olympos` and `ocean` - we try to use meaningful names :grinning:) running in **RAID 6** (by hardware) with **237 TB** each, and I would like to have **474 TB** in my _MinIO setup_, without losing any storage space and enabling further horizontal expansion. Considering that the `fs backend` was [ripped out](https://github.com/minio/minio/discussions/15967), I tried a multi-node setup using `xl-single backend`.

```yaml
version: '3.7'

services:
  s3:
    container_name: minio_s3
    hostname: ${HOSTNAME}
    image: minio/minio:RELEASE.2022-10-20T00-55-09Z
    command: server --console-address ":9001" http://olympos/data http://ocean/data
    restart: ${RESTART_POLICY}
    ports:
      - "9000:9000"
      - "9001:9001"
    extra_hosts:
      - "ocean:192.168.1.64"
      - "olympos:192.168.1.65"
    environment:
      - TZ
      - MINIO_ROOT_USER=${ACCESS_KEY}
      - MINIO_ROOT_PASSWORD=${SECRET_KEY}
    volumes:
      - /mnt/${HOSTNAME}/.minio/data:/data
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:9000/minio/health/live"]
      interval: 60s
      timeout: 60s
      retries: 3
```

And it's working fine.
```bash
$ mc admin info ocean
●  ocean:9000
   Uptime: 3 days 
   Version: 2022-10-20T00:55:09Z
   Network: 2/2 OK 
   Drives: 1/1 OK 
   Pool: 1

●  olympos:9000
   Uptime: 3 days 
   Version: 2022-10-20T00:55:09Z
   Network: 2/2 OK 
   Drives: 1/1 OK 
   Pool: 1

Pools:
   1st, Erasure sets: 1, Drives per erasure set: 2

17 TiB Used, 10 Buckets, 74,929 Objects
2 drives online, 0 drives offline
```

However after moving some data and check the usage space
```bash
$ df -h /mnt/ocean/ /mnt/olympos/
Filesystem                Size  Used Avail Use% Mounted on
//192.168.1.64/Ocean_w    237T   19T  218T   8% /mnt/ocean
//192.168.1.65/Olympos_w  237T   19T  218T   8% /mnt/olympos
```

... I'm not sure if the data is replicated or split on both servers. 

In a nutshell, I'm looking for a **Multi-Node Single-Drive** setup able to combine the storage space of `n` server without lose any storage space. I couldn't find it in the documentation / GitHub so I'm asking for help, thanks in advance.

**Comment by harshavardhana** [comment, marked as answer]
Data is going to be erasure coded with 2 disks so it is 1 data and 1 parity (think of it like replication) in this setup.  There is no HA in this setup you can lose 1 drive to have READ quorum however you have no WRITE quorum you need both servers online. 

This is not a recommended setup if you heed the warning that is displayed on the container logs - watch it via `docker logs`

**Comment by leandroleal** [comment]
Thank for the fast reply. 

What would be a setup for combing the storage space of the two servers without losing any storage space (RAID 0 / data striping only)? A multi-node setup using `fs backend` would support it?

**Reply by klauspost** [reply]
You shouldn't provide abstracted storage to MinIO. Provide raw disks as JBOD.

You can set `EC:0`, which will remove all redundancy. Use at your own risk, if you lose data you have been warned this is a bad approach.

**Reply by leandroleal** [reply]
Thank you for the reply and explanation. 

Each server has 15 x 20 Tb disks and I considered provided the raw disks as JBOB, however it would result in way less usable space than a RAID 6 + no redundancy in MinIO. Would this be a big issue for MinIO? 

I'm trying to find a balance between more usable space and fault tolerance.

**Reply by klauspost** [reply]
Just provide them as JBOD and MinIO will take care of the rest.

MinIO does erasure coding similar to RAID 6, so you don't lose more available space. The only difference is that your data is spread across both servers, also being faster and more resilient. By default minio will use 4 of 15 disks for erasure codes, which provides better safety than 2/15 in a RAID6. You can force MinIO to use 2/15, but that leaves little room for failures.

With 2 servers you cannot achieve failover, which is why we recommend starting with 4 servers for a production-grade setup. You can however set up 2 servers with 15 disks each as a test setup.

**Reply by leandroleal** [reply]
Great to know about! 
So, in summary, my options to optimize the available space (474 TB) using two server (15 x HDD 20TB each) are:

1. **Multi-Node Multi-Drive**: JBOD + MinIO erasure coding (`EC:2`)
2. **Multi-Node Single-Drive**: 
2.1. _RAID6 + MinIO erasure coding_ (`EC:0`  from `RELEASE.2022-10-29T06-21-33Z`)
2.2. _RAID6 + MinIO file system backend_ (before `RELEASE.2022-10-29T06-21-33Z`) 

I'm understanding that you firmily recommed `1`, however `2.x` would also be a valid option? 
I tend to think that `2.x` would provide more flexibility for further horizontal expansion, allowing me to expand the cluster using more heterogeneous servers and diferent set of HDD. Am I wrong here?

**Reply by ravindk89** [reply]
As with anything, you (and our community) can always apply more creative uses of MinIO in a variety of environments. 

2.x squarely falls into "Yes, that could work, and no we don't recommend it."

**Reply by Pacman-Clyde** [reply]
What about a multi node single drive set up with 4 rpi nodes with single drives attached to them? Would this work?

**Reply by ravindk89** [reply]
@Gasmanc same note as above - it would work, but with one drive per node, losing a drive and losing a node is indistinguishable. So you effectively have the same kind of failure tolerance - you can lose at most two disks and maintain reads, or two nodes and maintain reads. If you lose three drives or nodes (or any combination of the two), you have data loss.

For something like a personal HA server you can put into your closet, I'd actually suggest looking at boards like the ODroid H3+ or similar where there are onboard SATA ports - that way you can at least have two drives per "node". I think Pine64 has a board that has a PCI port that supports some SATA cards that might also be an option. These take more time, work, and resources - but you're also less likely to see issues because more than 2 of your rpis (or their USB drives) go down at once.
